### PR TITLE
enhancement(remap): Add VRL `match_array` function

### PR DESCRIPTION
**Pull request by leebenson**
Adds a `match_array` function, for matching a regex `pattern` against at least one element of an array `value`.

An optional `all:true` can be provided to require matching against all.

Motivated by a requirement surfaced in https://github.com/timberio/vector/pull/7633#discussion_r644143907 to parse wildcard patterns on elements inside of a _tags_ array.

Signed-off-by: Lee Benson <lee@leebenson.com>

### CODE REVIEW

File: lib/vrl/stdlib/src/lib.rs [hunk 1]
```diff
@@ -488,8 +492,10 @@ pub fn all() -> Vec<Box<dyn vrl::Function>> {
         Box::new(Log),
         #[cfg(feature = "match")]
         Box::new(Match),
-        #[cfg(feature = "match")]
+        #[cfg(feature = "match_any")]
```
Comment (leebenson) [inline lib/vrl/stdlib/src/lib.rs:495]: Not strictly related, but this seemed like a mismatch.

File: lib/vrl/stdlib/src/lib.rs [hunk 2]
```diff
@@ -550,8 +556,10 @@ pub fn all() -> Vec<Box<dyn vrl::Function>> {
         Box::new(Push),
         #[cfg(feature = "match")]
         Box::new(Match),
-        #[cfg(feature = "match")]
+        #[cfg(feature = "match_any")]
```
Comment (leebenson) [inline lib/vrl/stdlib/src/lib.rs:559]: Another mismatch.

File: docs/reference/remap/functions/match_array.cue [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package metadata
+
+remap: functions: match_array: {
+	category: "Enumerate"
+	description: """
+		Determines whether at least one element in the `value` array matches the `pattern`.
```
Comment (StephenWakely) [inline docs/reference/remap/functions/match_array.cue:6]: Maybe:

```suggestion
		Determines whether the elements in the `value` array matches the `pattern` - by default it checks at least one element matches, but can be set to determine if all the elements match.
```

**Review by StephenWakely — APPROVED** [review]
Looks great!

Would you be able to add a bench as well : https://github.com/timberio/vector/blob/master/lib/vrl/stdlib/benches/benches.rs

File: lib/vrl/stdlib/src/lib.rs (hunk 1 above)
Comment (StephenWakely) [reply lib/vrl/stdlib/src/lib.rs:495]: Yeah, good catch.

**Comment by leebenson** [comment]
@StephenWakely, does 0cd4ecd25c0a5c6d6a87ebae422b164e81753769 look reasonable for benches?

Not sure how to run them individually to bench a single function.

**Comment by leebenson** [comment]
(Bench addendum: d48ce333c4e87467e53c24e33e560d92e2e03ee5)

**Comment by leebenson** [comment]
Ah, `cargo bench match_array` -- I was trying `--bench`.

**Comment by StephenWakely** [comment]
> @StephenWakely, does [0cd4ecd](https://github.com/timberio/vector/commit/0cd4ecd25c0a5c6d6a87ebae422b164e81753769) look reasonable for benches?

Looks good to me!

**Review by jszwedko — COMMENTED** [review]
👍 this seems like a fine stopgap. In the future, I think we'd just support some sort of iteration / mapping to enable cases like this though. For example maybe we'd permit something like: `any(.some_array, |element| match(...))` and an analogous `all`.

**Pull request merged**